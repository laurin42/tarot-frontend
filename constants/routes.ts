import type { Route } from '@/types/navigation';

/**
 * Typsichere Routenkonstanten für die Navigation
 */
export const APP_ROUTES = {
  TABS: {
    THREECARDS: "/(tabs)/threecards" as Route,
    DAILYCARD: "/(tabs)/dailycard" as Route,
    PROFILE: "/(tabs)/profile" as Route,
  },
  AUTH: "/(auth)" as Route,
  HOME: "/" as Route,
};

/**
 * Konstanten für dynamische Routen
 */
export const DYNAMIC_ROUTES = {
  DEV_TOOLS: "/dev-tools" as const,
};