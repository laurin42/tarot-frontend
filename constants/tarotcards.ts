export interface ITarotCard {
  id: string;
  name: string;
  image: any;
}

export interface ISelectedAndShownCard extends ITarotCard {
  name: string;
  showFront?: boolean;
  isSelected?: boolean;
  currentIndex?: number;
  explanation?: string;
  image: any;
  onNextCard: () => void;
}

export const tarotCards: ITarotCard[] = [
  {
    id: "0",
    name:  "Der Narr",
    image: require("@/assets/images/tarot_cards/The_fool.jpg"),
  },
  {
    id: "1",
    name:  "Der Magier",
    image: require("@/assets/images/tarot_cards/The_magician.jpg"),
  },
  {
    id: "2",
    name: "Die Kaiserin",
    image: require("@/assets/images/tarot_cards/The_empress.jpg"),
  },
  {
    id: "3",
    name: "Die Hohepriesterin",
    image: require("@/assets/images/tarot_cards/High_priestess.jpg"),
  },
  {
    id: "4",
    name: "König der Kelche",
    image: require("@/assets/images/tarot_cards/King_of_cups.jpg"),
  },
  {
    id: "5",
    name: "Ritter der Kelche",
    image: require("@/assets/images/tarot_cards/Knight_of_cups.jpg"),
  },
  {
    id: "6",
    name: "Ritter der Münzen",
    image: require("@/assets/images/tarot_cards/Knight_of_pentacles.jpg"),
  },
  {
    id: "7",
    name: "Page der Kelche",
    image: require("@/assets/images/tarot_cards/Page_of_cups.jpg"),
  },
  {
    id: "8",
    name: "Page der Münzen",
    image: require("@/assets/images/tarot_cards/Page_of_pentacles.jpg"),
  },
  {
    id: "9",
    name: "Königin der Kelche",
    image: require("@/assets/images/tarot_cards/Queen_of_cups.jpg"),
  },
  {
    id: "10",
    name: "Zwei Kelche",
    image: require("@/assets/images/tarot_cards/Two_Cups.jpg"),
  },
  {
    id: "11",
    name: "Drei Kelche",
    image: require("@/assets/images/tarot_cards/Three_Cups.jpg"),
  },
  {
    id: "12",
    name: "Vier Kelche",
    image: require("@/assets/images/tarot_cards/Four_Cups.jpg"),
  },
  {
    id: "13",
    name: "Fünf Kelche",
    image: require("@/assets/images/tarot_cards/Five_Cups.jpg"),
  },
  {
    id: "14",
    name: "Sechs Kelche",
    image: require("@/assets/images/tarot_cards/Six_Cups.jpg"),
  },
  {
    id: "15",
    name: "Sieben Kelche",
    image: require("@/assets/images/tarot_cards/Seven_Cups.jpg"),
  },
  {
    id: "16",
    name: "Acht Kelche",
    image: require("@/assets/images/tarot_cards/Eight_Cups.jpg"),
  },
  {
    id: "17",
    name: "Neun Kelche",
    image: require("@/assets/images/tarot_cards/Nine_Cups.jpg"),
  },
  {
    id: "18",
    name: "Zehn Kelche",
    image: require("@/assets/images/tarot_cards/Ten_Cups.jpg"),
  },
  {
    id: "19",
    name: "Drei Münzen",
    image: require("@/assets/images/tarot_cards/Three_pents.jpg"),
  },
  {
    id: "20",
    name: "Vier Münzen",
    image: require("@/assets/images/tarot_cards/Four_pents.jpg"),
  },
  {
    id: "21",
    name: "Fünf Münzen",
    image: require("@/assets/images/tarot_cards/Five_pents.jpg"),
  },
  {
    id: "22",
    name: "Sechs Münzen",
    image: require("@/assets/images/tarot_cards/Six_pents.jpg"),
  },
  {
    id: "23",
    name: "Sieben Münzen",
    image: require("@/assets/images/tarot_cards/Seven_pents.jpg"),
  },
  {
    id: "24",
    name: "Acht Münzen",
    image: require("@/assets/images/tarot_cards/Eight_pents.jpg"),
  },
  {
    id: "25",
    name: "Neun Münzen",
    image: require("@/assets/images/tarot_cards/Nine_pents.jpg"),
  },
  {
    id: "26",
    name: "Zehn Münzen",
    image: require("@/assets/images/tarot_cards/Ten_pents.jpg"),
  },
  {
    id: "27",
    name: "Ass Münzen",
    image: require("@/assets/images/tarot_cards/Ace_of_Pentacles.jpg"),
  },
];


