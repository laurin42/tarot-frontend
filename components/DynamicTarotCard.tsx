import React, { useState, useEffect } from "react";
import { View, StyleSheet, Text } from "react-native";
import { Image } from "expo-image";
import { getCardImageByName, CARD_BACK_IMAGES } from "@/constants/tarotcards";

interface DynamicTarotCardProps {
  cardName: string; // Name der Karte, wird zum dynamischen Laden verwendet
  isShown?: boolean; // Ob Vorder- oder Rückseite gezeigt wird
  cardBackIndex?: number; // Index für die gewünschte Kartenrückseite (0, 1, oder 2)
  style?: any; // Zusätzliche Styles (will now control size)
}

/**
 * Dynamische Tarotkarten-Komponente, die Bilder basierend auf dem Namen automatisch lädt
 *
 * Verwendung:
 * <DynamicTarotCard
 *   cardName="Der Narr"
 *   size="medium"
 *   isShown={true}
 * />
 */
export default function DynamicTarotCard({
  cardName,
  isShown = true,
  cardBackIndex = 0, // Standardmäßig die erste Rückseite (Index 0)
  style,
}: DynamicTarotCardProps) {
  const [imageSource, setImageSource] = useState<any>(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    // Stelle sicher, dass der Index gültig ist
    const validCardBackIndex =
      cardBackIndex >= 0 && cardBackIndex < CARD_BACK_IMAGES.length
        ? cardBackIndex
        : 0;

    if (isShown) {
      try {
        const source = getCardImageByName(cardName);
        setImageSource(source);
        setIsLoading(false);
      } catch (err) {
        console.error("Fehler beim Laden der Karte:", err);
        setError("Konnte Karte nicht laden");
        setIsLoading(false);
      }
    } else {
      // Wenn die Rückseite gezeigt werden soll, lade das entsprechende Bild aus dem Array
      setImageSource(CARD_BACK_IMAGES[validCardBackIndex]);
      setIsLoading(false);
      setError(null); // Setze Fehler zurück, falls zuvor einer aufgetreten ist
    }
  }, [cardName, isShown, cardBackIndex]); // Füge cardBackIndex zur Abhängigkeitsliste hinzu

  // Wähle die korrekte Bildquelle basierend auf isShown und cardBackIndex
  const finalImageSource = isShown
    ? imageSource
    : CARD_BACK_IMAGES[
        cardBackIndex >= 0 && cardBackIndex < CARD_BACK_IMAGES.length
          ? cardBackIndex
          : 0
      ];

  return (
    <View style={[styles.container, style]}>
      {isLoading ? (
        <Text style={styles.loadingText}>Loading...</Text>
      ) : error ? (
        <View style={styles.errorContainer}>
          <Text style={styles.errorText}>{error}</Text>
        </View>
      ) : (
        <Image
          style={styles.image}
          // Verwende die ermittelte Bildquelle
          source={finalImageSource}
          // placeholder={isShown ? blurhash : undefined} // Beispielhaft auskommentiert, falls du Placeholder verwendest
          contentFit="contain"
          transition={200}
        />
      )}
      {isShown && (
        <View style={styles.nameContainer}>
          <Text style={styles.nameText}>{cardName}</Text>
        </View>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    position: "relative",
    borderWidth: 0.3,
    borderColor: "black", // Testweise auf schwarz ändern
    borderRadius: 16,
    overflow: "hidden",
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#000000", // Test: Undurchsichtiger schwarzer Hintergrund
  },
  image: {
    width: "100%",
    height: "100%",
  },
  nameContainer: {
    position: "absolute",
    bottom: 4,
    left: 4,
    right: 4, // Sorge dafür, dass der Text ggf. umbricht
    backgroundColor: "rgba(0, 0, 0, 0.6)", // Etwas dunkler für besseren Kontrast
    borderRadius: 4,
    paddingVertical: 2,
    paddingHorizontal: 4, // Etwas mehr Padding
  },
  nameText: {
    color: "white",
    fontSize: 12, // Kleinere Schriftgröße für Namen
    fontWeight: "bold",
    textAlign: "center", // Zentrierter Text
  },
  loadingText: {
    color: "rgba(255, 215, 0, 0.8)", // Goldene Ladeanzeige
    textAlign: "center",
    padding: 10,
  },
  errorContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    padding: 10,
    backgroundColor: "rgba(255, 0, 0, 0.1)", // Leichter roter Hintergrund bei Fehler
  },
  errorText: {
    color: "red",
    textAlign: "center",
    fontSize: 12,
  },
});
