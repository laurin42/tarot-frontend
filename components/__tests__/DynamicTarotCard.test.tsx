import React from "react";
import { render, screen, waitFor } from "@testing-library/react-native";
import DynamicTarotCard from "../DynamicTarotCard"; // Pfad anpassen
import { getCardImageByName } from "../../constants/tarotCards";

// --- Mock-Daten (können hier bleiben) ---
const mockCardBackSource1 = "mock-card-back-1";
const mockCardBackSource2 = "mock-card-back-2";
const mockKnownCardSource = "mock-known-card-source";
const mockFallbackSource = mockCardBackSource1;

// --- Expliziter Mock mit relativem Pfad ---
jest.mock("../../constants/tarotcards", () => ({
  // Wichtig: __esModule: true, wenn das Originalmodul ES Modules verwendet (wahrscheinlich)
  __esModule: true,
  getCardImageByName: jest.fn((cardName: string) => {
    if (cardName === "Known Card") {
      return mockKnownCardSource;
    }
    console.warn(`MOCK: Keine Bildabbildung für Karte "${cardName}" gefunden`);
    return mockFallbackSource;
  }),
  CARD_BACK_IMAGES: [mockCardBackSource1, mockCardBackSource2],
  // Füge hier leere Mocks für andere Exporte hinzu, falls nötig
  tarotCards: [],
  ITarotCard: {},
  ISelectedAndShownCard: {},
}));

// --- Mock expo-image ---
// Mock implementation renders a View and passes props through for snapshot verification
jest.mock("expo-image", () => ({
  Image: jest.fn(({ style, source, contentFit, transition, ...rest }) => {
    const ReactActual = jest.requireActual("react");
    const ViewActual = jest.requireActual("react-native").View;
    // Include testID or other identifiable props based on source/style if needed
    // For simplicity, we pass source and style directly as props to the View
    return ReactActual.createElement(ViewActual, {
      style: style,
      testID: "mock-expo-image",
      "data-source": source, // Use data- attributes for custom props in snapshots
      "data-contentFit": contentFit,
      "data-transition": transition,
      ...rest,
    });
  }),
}));

// Cast zu jest.Mock für Typ-Sicherheit
const mockGetCardImageByNameImpl = getCardImageByName as jest.Mock;

// Mock console.warn
let consoleWarnSpy: jest.SpyInstance;

describe("DynamicTarotCard Component", () => {
  beforeEach(() => {
    // Verwende die gecastete Mock-Variable
    mockGetCardImageByNameImpl.mockClear();
    consoleWarnSpy = jest.spyOn(console, "warn").mockImplementation(() => {});
  });

  afterEach(() => {
    consoleWarnSpy.mockRestore();
    // Kein Spy mehr auf getCardImageByName nötig, da wir den Mock direkt haben
  });

  it("renders front side correctly for a known card", async () => {
    render(<DynamicTarotCard cardName="Known Card" isShown={true} />);
    await waitFor(() => {
      expect(mockGetCardImageByNameImpl).toHaveBeenCalledWith("Known Card");
    });
    const tree = screen.toJSON();
    expect(tree).toMatchSnapshot();
    const image = screen.getByTestId("mock-expo-image");
    // Erwarte den einfachen Mock-String
    expect(image.props["data-source"]).toBe(mockKnownCardSource);
    expect(screen.getByText("Known Card")).toBeTruthy();
    expect(consoleWarnSpy).not.toHaveBeenCalled();
  });

  it("renders back side correctly with default index (0)", async () => {
    render(<DynamicTarotCard cardName="Any Card" isShown={false} />);
    await new Promise((res) => setImmediate(res));
    const tree = screen.toJSON();
    expect(tree).toMatchSnapshot();
    const image = screen.getByTestId("mock-expo-image");
    expect(mockGetCardImageByNameImpl).not.toHaveBeenCalled();
    expect(screen.queryByText("Any Card")).toBeNull();
  });

  it("renders back side correctly with specific index (1)", async () => {
    render(
      <DynamicTarotCard cardName="Any Card" isShown={false} cardBackIndex={1} />
    );
    await new Promise((res) => setImmediate(res));
    const tree = screen.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders back side with default index (0) if provided index is invalid", async () => {
    render(
      <DynamicTarotCard
        cardName="Any Card"
        isShown={false}
        cardBackIndex={99}
      />
    );
    await new Promise((res) => setImmediate(res));
    const tree = screen.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders fallback image (first back) for an unknown card when shown", async () => {
    render(<DynamicTarotCard cardName="Unknown Card" isShown={true} />);
    await waitFor(() => {
      expect(mockGetCardImageByNameImpl).toHaveBeenCalledWith("Unknown Card");
    });
    const tree = screen.toJSON();
    expect(tree).toMatchSnapshot();
    const image = screen.getByTestId("mock-expo-image");
    // Erwarte den einfachen Mock-String (Fallback)
    expect(image.props["data-source"]).toBe(mockFallbackSource);
    expect(screen.getByText("Unknown Card")).toBeTruthy();
    // Prüfe die Warnung aus der Mock-Implementierung
    expect(consoleWarnSpy).toHaveBeenCalledWith(
      'MOCK: Keine Bildabbildung für Karte "Unknown Card" gefunden'
    );
  });

  it("applies custom style correctly", async () => {
    const customStyle = { width: 200, height: 300, opacity: 0.5 };
    render(
      <DynamicTarotCard
        cardName="Known Card"
        isShown={true}
        style={customStyle}
      />
    );
    await waitFor(() => {
      expect(mockGetCardImageByNameImpl).toHaveBeenCalledWith("Known Card");
    });
    const tree = screen.toJSON();
    expect(tree).toMatchSnapshot();

    const stylesArray =
      tree && !Array.isArray(tree) && Array.isArray(tree.props.style)
        ? tree.props.style
        : [];
    expect(stylesArray).toEqual(
      expect.arrayContaining([expect.objectContaining(customStyle)])
    );
  });

  // Add tests for loading and error states if possible/needed
  // This might require more complex mocking of the useEffect or getCardImageByName
});
