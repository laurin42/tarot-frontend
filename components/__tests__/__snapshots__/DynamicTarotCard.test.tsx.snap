// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicTarotCard Component applies custom style correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#000000",
        "borderColor": "black",
        "borderRadius": 16,
        "borderWidth": 0.3,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
      },
      {
        "height": 300,
        "opacity": 0.5,
        "width": 200,
      },
    ]
  }
>
  <View
    data-contentFit="contain"
    data-source="mock-known-card-source"
    data-transition={200}
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
    testID="mock-expo-image"
  />
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.6)",
        "borderRadius": 4,
        "bottom": 4,
        "left": 4,
        "paddingHorizontal": 4,
        "paddingVertical": 2,
        "position": "absolute",
        "right": 4,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 12,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Known Card
    </Text>
  </View>
</View>
`;

exports[`DynamicTarotCard Component renders back side correctly with default index (0) 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#000000",
        "borderColor": "black",
        "borderRadius": 16,
        "borderWidth": 0.3,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
      },
      undefined,
    ]
  }
>
  <View
    data-contentFit="contain"
    data-transition={200}
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
    testID="mock-expo-image"
  />
</View>
`;

exports[`DynamicTarotCard Component renders back side correctly with specific index (1) 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#000000",
        "borderColor": "black",
        "borderRadius": 16,
        "borderWidth": 0.3,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
      },
      undefined,
    ]
  }
>
  <View
    data-contentFit="contain"
    data-transition={200}
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
    testID="mock-expo-image"
  />
</View>
`;

exports[`DynamicTarotCard Component renders back side with default index (0) if provided index is invalid 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#000000",
        "borderColor": "black",
        "borderRadius": 16,
        "borderWidth": 0.3,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
      },
      undefined,
    ]
  }
>
  <View
    data-contentFit="contain"
    data-transition={200}
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
    testID="mock-expo-image"
  />
</View>
`;

exports[`DynamicTarotCard Component renders fallback image (first back) for an unknown card when shown 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#000000",
        "borderColor": "black",
        "borderRadius": 16,
        "borderWidth": 0.3,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
      },
      undefined,
    ]
  }
>
  <View
    data-contentFit="contain"
    data-source="mock-card-back-1"
    data-transition={200}
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
    testID="mock-expo-image"
  />
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.6)",
        "borderRadius": 4,
        "bottom": 4,
        "left": 4,
        "paddingHorizontal": 4,
        "paddingVertical": 2,
        "position": "absolute",
        "right": 4,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 12,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Unknown Card
    </Text>
  </View>
</View>
`;

exports[`DynamicTarotCard Component renders front side correctly for a known card 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#000000",
        "borderColor": "black",
        "borderRadius": 16,
        "borderWidth": 0.3,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
      },
      undefined,
    ]
  }
>
  <View
    data-contentFit="contain"
    data-source="mock-known-card-source"
    data-transition={200}
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
    testID="mock-expo-image"
  />
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.6)",
        "borderRadius": 4,
        "bottom": 4,
        "left": 4,
        "paddingHorizontal": 4,
        "paddingVertical": 2,
        "position": "absolute",
        "right": 4,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 12,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Known Card
    </Text>
  </View>
</View>
`;
